File: backend/asterion_api/harness.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any, Literal, Mapping

PrivacyLevel = Literal["local", "hybrid", "external"]


class BaseHarness(ABC):
    """Common optimization interface for Meta-Harness candidate services."""

    privacy_level: PrivacyLevel = "local"

    @abstractmethod
    async def execute(self, payload: Mapping[str, Any] | None = None) -> Any:
        """Run the service's default harness action."""

    @abstractmethod
    def get_state(self) -> dict[str, Any]:
        """Return serializable service state for harness inspection."""

    @abstractmethod
    def set_state(self, state: Mapping[str, Any]) -> None:
        """Restore service state from a previous harness candidate."""

File: crates/core/src/harness.rs
use std::collections::HashMap;

use serde_json::Value;

/// BaseHarness trait — every service must implement this.
/// Mirrors `asterion_api.harness.BaseHarness` from Python.
pub trait BaseHarness: Send + Sync {
    fn privacy_level(&self) -> &str;

    fn execute(&self, payload: Option<HashMap<String, Value>>) -> Value;

    fn get_state(&self) -> HashMap<String, Value>;

    fn set_state(&self, state: HashMap<String, Value>);
}
